### PR TITLE
Apply constant propagation in the TypeChecker

### PR DESCRIPTION
**Pull request by leonardoalt**
For `BinaryOperation` visits.

I'm not sure about this PR... the error messages looked much better before. Even if this is done only for constant variable declarations, it's still worse (some of the tests have that case).

### CODE REVIEW

**Comment by chriseth** [comment]
I think the `ConstStateVarCircularReferenceChecker` can run much earlier, that would bring back the nice error messages.

**Comment by chriseth** [comment]
Oh wow, the external tests saved us from a regression here:
```
Error: CompileError: /tmp/tmp.bDbAuAngYc/ext/contracts/ReputationMiningCycle.sol:79:24: TypeError: Type rational_const 7719...(68 digits omitted)...9329 / 5760 is not implicitly convertible to expected type uint256. Try converting to type ufixed256x4 or use an explicit conversion.
  uint256 constant X = UINT256_MAX / MINING_WINDOW_SIZE;
                       ^------------------------------^
```
Can you add a test like that?
The bug might even be present in develop:
What does f return?
```
uint constant a = 12;
uint constant b = 10;
contract C {
  function f() public pure returns (uint) {
    uint[(a/b)*b] memory x;
    return x.length;
  }
}
```

**Comment by leonardoalt** [comment]
Indeed `f` returns 12.

File: libsolidity/analysis/TypeChecker.cpp [hunk 1]
```diff
@@ -1565,10 +1568,15 @@ void TypeChecker::endVisit(BinaryOperation const& _operation)
...
-		TokenTraits::isCompareOp(_operation.getOperator()) ?
-		TypeProvider::boolean() :
-		commonType;
+	if (TokenTraits::isCompareOp(_operation.getOperator()))
+		_operation.annotation().type = TypeProvider::boolean();
+	else
+	{
+		_operation.annotation().type = commonType;
+		if (originalCommonType)
+			if (TypePointer constantType = ConstantEvaluator(m_errorReporter).evaluate(_operation))
```
Comment (chriseth) [inline libsolidity/analysis/TypeChecker.cpp:1577]: Does this handle forward and backward references well?

**Comment by leonardoalt** [comment]
I'm closing this PR since I'm not sure it really makes sense and is safe.

**Pull request closed without merging**